src/com/twitter/intellij/pants/ui/PantsToBspProjectAction.java

```diff
@@ -0,0 +1,192 @@
+    ProgressManager.getInstance().run(new Task.Backgroundable(project, "Preparing BSP Project", false) {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        try {
+          GeneralCommandLine commandLine = createCommandLine(project);
+          Process process = commandLine.createProcess();
+          String output = readToString(process.getInputStream());
+          String outputErr = readToString(process.getErrorStream());
+          int result = process.waitFor();
+          if (result != 0) {
+            Path workspace = commandLine.getWorkDirectory().toPath();
+            Optional<Path> projectPath = existingProjectPath(outputErr, workspace);
+
+            projectPath.ifPresent(path -> registerNewBspProjectAndOpen(path, project));
+            projectPath.orElseThrow(() -> new RuntimeException(formatError(output, outputErr)));
+          } else {
+            Path projectPath = Paths.get(output).toAbsolutePath();
+            registerNewBspProjectAndOpen(projectPath, project);
+          }
+        }
+        catch (Exception ex) {
+          ApplicationManager.getApplication()
+            .invokeLater(() -> Messages.showErrorDialog(project, ex.getMessage(), "Error"));
+        }
+      }
+    });
+  }
+
+  private void registerNewBspProjectAndOpen(Path projectPath, Project project) {
+    PropertiesComponent.getInstance(project).setValue(BSP_LINKED_PROJECT_PATH, projectPath.toString());
+    ApplicationManager.getApplication()
+      .invokeLater(() -> ProjectUtil.openOrImport(projectPath, new OpenProjectTask()));
+  }
+
+  private Optional<Path> existingProjectPath(String output, Path workspace) {
+    Pattern pattern = Pattern.compile("can't create project named '(.*?)' because it already exists");
+    Matcher matcher = pattern.matcher(output);
+    if (matcher.find()) {
+      String name = matcher.group(1);
+      return Optional.of(workspace.resolveSibling("bsp-projects").resolve(name));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private GeneralCommandLine createCommandLine(Project project) throws IOException {
+    GeneralCommandLine commandLine = PantsUtil.defaultCommandLine(project);
+
+    String coursier = coursierPath().toString();
+    commandLine.setExePath(coursier);
+
+    List<String> commandBase = Arrays.asList(
+      "launch", "org.scalameta:metals_2.12:latest.stable",
+      "--main", "scala.meta.internal.pantsbuild.BloopPants",
+      "--",
+      "create",
+      "--intellij",
+      "--intellijLauncher", "echo"
+    );
+    commandLine.addParameters(commandBase);
+
+    List<String> targets = pantsTargets(project);
+    commandLine.addParameters(targets);
+    return commandLine;
+  }
+
+  private Path coursierPath() throws IOException {
+    Path destination = Paths.get(System.getProperty("java.io.tmpdir"), "pants-plugin-coursier");
+    if (!Files.exists(destination)) {
+      URL url = new URL("https://git.io/coursier-cli");
```

Comment:
We download newest coursier here. Imo this should either be changed to  some fixed version, or logged. Without it we lose information about coursier version that may be essential to reproduce potential bugs.

Reply:
This is a good point. If we copy the coursier bootstrap script from resources then we always use the same coursier version